File: app/src/main/java/com/alwaysbaked/population/MainActivity.java
package com.alwaysbaked.population;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;

import com.alwaysbaked.population.Model.Root;
import com.alwaysbaked.population.Retrofit.API;
import com.alwaysbaked.population.Retrofit.RetrofitClient;
import com.alwaysbaked.population.Util.Adapter;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";

    private API myAPI;
    private CompositeDisposable disposable = new CompositeDisposable();
    private Retrofit retrofit;

    @BindView(R.id.rvFeed)
    RecyclerView mFeed;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.d(TAG, "onCreate: started.");
        ButterKnife.bind(this);

        //initialise
        retrofit = RetrofitClient.getInstance();
        myAPI = retrofit.create(API.class);

        //recycler view
        mFeed.setHasFixedSize(true);
        mFeed.setLayoutManager(new LinearLayoutManager(this));

        fetchData();
    }

    private void fetchData() {
        disposable.add(myAPI.getTheWorldPopulation()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<Root>() {
                    @Override
                    public void accept(Root root) throws Exception {
                        displayData(root);
                    }
                }));

    }

    private void displayData(Root root) {
        Adapter adapter = new Adapter(this, root.getWorldpopulation());
        mFeed.setAdapter(adapter);
    }

    @Override
    protected void onStop() {
        disposable.clear();
        super.onStop();
    }
}

File: app/src/main/java/com/alwaysbaked/population/MainActivity.kt
package com.alwaysbaked.population

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import com.alwaysbaked.population.model.Root
import com.alwaysbaked.population.retrofit.APIService
import com.alwaysbaked.population.util.Adapter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "Main2Activity"
    }

    private val apiServe by lazy {
        APIService.create()
    }

    private var disposable: Disposable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Log.d(TAG, "onCreate: started.")

        //recycler view
        rvFeed.layoutManager = LinearLayoutManager(this)

        fetch()

    }

    private fun fetch() {
        Log.d(TAG, "fetch: started.")
        disposable = apiServe.theWorldPopulation
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { root ->
                    displayData(root)
                }

    }

    private fun displayData(root: Root) {
        val adapter = Adapter(this, root.worldpopulation)
        rvFeed.adapter = adapter
    }

    override fun onPause() {
        super.onPause()
        disposable?.dispose()
    }
}
